### PR TITLE
Add flexible unet for encoder-decoder structure

### LINKED ISSUE #4533: Flexible UNet implementation with backbones
**Is your feature request related to a problem? Please describe.**
Followup discussion of #896 .
Currently, we don't have a flexible segmentation pipeline that can support different backbones and pretrained weights.
I think it would be very useful to provide such networks, at least for the UNet implementation.
Actually, we already have some similar prototype in the tutorial:
https://github.com/Project-MONAI/tutorials/blob/main/kaggle/RANZCR/4th_place_solution/models/seg_model.py
I think we can start from integrating this implementation.
@wyli @yiheng-wang-nv What do you think?

### PR DESCRIPTION
**Pull request by binliunls**
A flexible U-Net implements an encoder-decoder network. This network uses U-Net
as decoder and efficientnet-V1 family as encoder. The encoder type can be set
by passing parameters 'efficientnet-b0, ..., efficientnet-b7, efficientnet-l2'
to the initialize function.

Signed-off-by: binliu <binliu@nvidia.com>

Fixes #4533  .

### Description
Add a new encoder-decoder defined network structure to monai. Currently the script is
the very beginning version of it and can only support U-Net as decoder and efficientnet-
V1 family as encoder.
Aim to make this more flexible and add more networks to it.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/networks/nets/flexible_unet.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        if act is not None:
+            act_layer = get_act_layer(act)
+        else:
+            act_layer = nn.Identity()
+        super().__init__(conv_layer, up_layer, act_layer)
+
+
+class FlexibleUNet(nn.Module):
+    """
+    A flexible implement of UNet, in which the backbone can be replaced with any efficient network.
```
Comment (yiheng-wang-nv) [inline monai/networks/nets/flexible_unet.py:191]: a flexible implementation
efficientnet

File: monai/networks/nets/flexible_unet.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+        self.backbone = backbone
+        self.spatial_dims = spatial_dims
+        model_name = backbone
+        encoder_channels = _get_encoder_channels_by_backbone(backbone, in_channels)
+        self.encoder = EfficientNetBNFeatures(
+            model_name=model_name,
+            pretrained=pretrained,
+            in_channels=in_channels,
+            norm=("batch", {"eps": 1e-3, "momentum": 0.1}),
```
Comment (yiheng-wang-nv) [inline monai/networks/nets/flexible_unet.py:226]: Hi @binliunls , here could the parameter of `norm` be an argument, thus users can adjust it. Same for `act` and `norm` of decoder. You can refer to: https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/nets/autoencoder.py#L107

File: monai/networks/nets/flexible_unet.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    Get the encoder output channels by given backbone name.
+
+    Args:
+        backbone: name of backbone to generate features, can be from [efficientnet-b0, ..., efficientnet-b7].
+        in_channels: channel of input tensor, default to 3.
+
+    Returns:
+        A tuple of output feature map channels' length .
+    """
+    encoder_channel_tuple = efficientnet_feature_channel[backbone]
```
Comment (yiheng-wang-nv) [inline monai/networks/nets/flexible_unet.py:51]: this function may expand to support more backbones, thus `efficientnet_feature_channel` can be replaced with an arg

File: monai/networks/nets/flexible_unet.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/networks/nets/flexible_unet.py:191]: Hi @yiheng-wang-nv ,

I think maybe @binliunls means efficient backbones not specific to `EfficientNet`?
Is this encoder-decoder pipeline possible to support other networks in MONAI like `DenseNet` and `SENet`?
If not, I think it's OK to at least support different EfficientNet for the first version.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @binliunls ,

I think the error of CI is due to the image size in the tests.
Please use smaller input size < 100, like 32, 16. Don't use 128, 256 or 224, etc.

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Looks good to me as the first version implementation.
We can add more backbone support in the following PRs.

Thanks.

File: monai/networks/nets/flexible_unet.py (hunk 1 above)
Comment (binliunls) [reply monai/networks/nets/flexible_unet.py]: Yes Nic. I will add more backbones to the second version implementation. And the name should remain.

File: monai/networks/nets/flexible_unet.py (hunk 2 above)
Comment (binliunls) [reply monai/networks/nets/flexible_unet.py]: Hi, @yiheng-wang-nv ，
    I have modified the code to make norm, active and other parameters to be an argument so that users can change them according to their demand.

File: monai/networks/nets/flexible_unet.py (hunk 3 above)
Comment (binliunls) [reply monai/networks/nets/flexible_unet.py]: Thanks for your advice @yiheng-wang-nv . I have modified this name.

File: tests/test_flexible_unet.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        batches=[1],
+        pretrained=[False],
+        in_channels=3,
+        num_classes=10,
+        input_shape=64,
+    )
+)
+
+
+class TestFLEXIBLEUNET(unittest.TestCase):
```
Comment (yiheng-wang-nv) [inline tests/test_flexible_unet.py:182]: HI @binliunls , this PR looks good to me, could you add one more test case which tests the pretrained weights for 2d input?
You can refer to `test_kitty_pretrained` in `test_efficientnet.py`

Comment (binliunls) [reply tests/test_flexible_unet.py:205]: Hi @yiheng-wang-nv ,
 I have added the weight check to the unitest.
Thanks.

**Review by yiheng-wang-nv — APPROVED** [review]

**Comment by Nic-Ma** [comment]
Looks good to me now.
@wyli @ericspod , if you don't have other comments, I will merge this PR soon as it's P0 task for us.

Thanks in advance.

File: tests/test_flexible_unet.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+            "pretrained": True,
+            "spatial_dims": 2,
+            "norm": ("batch", {"eps": 1e-3, "momentum": 0.01}),
+        },
+        ["_conv_stem.weight"],
+    )
+]
+
+
+class TestFLEXIBLEUNET(unittest.TestCase):
```
Comment (wyli) [inline tests/test_flexible_unet.py:205]: could you help add skip_if_quick for the entire class https://github.com/Project-MONAI/MONAI/blob/6da25ae54996a7be8e06b9e23a18a25ea44d70f2/tests/test_efficientnet.py#L283

File: monai/networks/nets/flexible_unet.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+
+
+class UNetDecoder(nn.Module):
+    """
+    UNet Decoder.
+    This class refers to `segmentation_models.pytorch
+    <https://github.com/qubvel/segmentation_models.pytorch>`_.
+
+    Args:
+        dim: number of spatial dimensions.
```
Comment (wyli) [inline monai/networks/nets/flexible_unet.py:64]: this argument should be `spatial_dims` instead of `dim`

File: monai/networks/nets/flexible_unet.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+
+    def __init__(
+        self,
+        spatial_dims: int,
+        encoder_channels: Sequence[int],
+        decoder_channels: Sequence[int],
+        act: Union[str, tuple],
+        norm: Union[str, tuple],
+        dropout: Union[float, tuple],
+        bias: bool,
```
Comment (wyli) [inline monai/networks/nets/flexible_unet.py:91]: missing the docstring for `bias`

File: monai/networks/nets/flexible_unet.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+        if act is not None:
+            act_layer = get_act_layer(act)
+        else:
+            act_layer = nn.Identity()
+        super().__init__(conv_layer, up_layer, act_layer)
+
+
+class FlexibleUNet(nn.Module):
+    """
+    A flexible implement of UNet.
```
Comment (wyli) [inline monai/networks/nets/flexible_unet.py:191]:
```suggestion
    A flexible implement of UNet-like encoder-decoder architecture.
```

File: monai/networks/nets/flexible_unet.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+                    pre_conv=pre_conv,
+                    interp_mode=interp_mode,
+                    align_corners=align_corners,
+                    halves=halve,
+                )
+            )
+        self.blocks = nn.ModuleList(blocks)
+
+    def forward(self, features: Sequence[torch.Tensor], skip_connect: int = 4):
+        skips = features[:-1][::-1]
```
Comment (wyli) [inline monai/networks/nets/flexible_unet.py:130]: why not adding the resizing like https://github.com/qubvel/segmentation_models.pytorch/blob/792c2735aa0e926afa17d1105eace63ef2dcc624/segmentation_models_pytorch/decoders/unet/decoder.py#L36
so that we dont have this constraint https://github.com/Project-MONAI/MONAI/blob/e82d15d9c3799b2dbe6a6960cc9e47fb2359e8f7/monai/networks/nets/flexible_unet.py#L297-L300

File: tests/test_flexible_unet.py (hunk 5 above)
Comment (Nic-Ma) [reply tests/test_flexible_unet.py:205]: Hi @wyli ,

Should we add the decorator to the entire class or only the `test_pretrain` function?

Thanks.

Comment (wyli) [reply tests/test_flexible_unet.py:205]: The class, as the downloading might be slow and causes timeout for the premerge ci

File: monai/networks/nets/flexible_unet.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+        super().__init__()
+
+        if backbone not in encoder_feature_channel:
+            raise ValueError(f"invalid model_name {backbone} found, must be one of {encoder_feature_channel.keys()}.")
+
+        if spatial_dims not in (2, 3):
+            raise ValueError("spatial_dims can only be 2 or 3.")
+
+        if "ap" in backbone:
+            adv_prop = True
```
Comment (wyli) [inline monai/networks/nets/flexible_unet.py:248]: These could be simplified: adv_drop = "ap" in backbine

File: monai/networks/nets/flexible_unet.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+            interp_mode=interp_mode,
+            pre_conv=None,
+            align_corners=None,
+        )
+
+        self.segmentation_head = SegmentationHead(
+            dim=spatial_dims, in_channels=decoder_channels[-1], out_channels=out_channels, kernel_size=3, act=None
+        )
+
+        self.pool = Pool[Pool.ADAPTIVEMAX, 2](1)
```
Comment (wyli) [inline monai/networks/nets/flexible_unet.py:282]: It's hardcoded for 2d? If it's a mistake,  Could you please add test cases after correct it

File: tests/test_flexible_unet.py (hunk 5 above)
Comment (binliunls) [reply tests/test_flexible_unet.py:206]: Thanks @wyli , I have added the @skip_if_quick to the whole class.

File: monai/networks/nets/flexible_unet.py (hunk 6 above)
Comment (binliunls) [reply monai/networks/nets/flexible_unet.py]: Thanks @wyli , have changed the parameter.

File: monai/networks/nets/flexible_unet.py (hunk 7 above)
Comment (binliunls) [reply monai/networks/nets/flexible_unet.py:92]: Thanks @wyli , the docstring for bias has been added.

File: monai/networks/nets/flexible_unet.py (hunk 8 above)
Comment (binliunls) [reply monai/networks/nets/flexible_unet.py]: Hi, @wyli , this issue has been modifed.

File: monai/networks/nets/flexible_unet.py (hunk 10 above)
Comment (binliunls) [reply monai/networks/nets/flexible_unet.py]: Hi @wyli ，have fixed this issue.

File: monai/networks/nets/flexible_unet.py (hunk 11 above)
Comment (binliunls) [reply monai/networks/nets/flexible_unet.py]: Sorry, it's a useless legacy code and won't influence the class function. I have deleted it .

File: monai/networks/nets/flexible_unet.py (hunk 9 above)
Comment (binliunls) [reply monai/networks/nets/flexible_unet.py:134]: Hi @wyli ,
 I tried to fix this by adding a is_pad parameter in a flexible unet class. Could you please help me to check if this change is ok.

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me. I'll merge this pr if all tests work fine.

**Comment by ibro45** [comment]
https://github.com/Project-MONAI/MONAI/discussions/6562
